crates/services/p2p/src/codecs/request_response.rs

```diff
@@ -0,0 +1,150 @@
+use std::io;
+
+use crate::request_response::{
+    messages::{
+        RequestMessage,
+        V1ResponseMessage,
+        V2ResponseMessage,
+    },
+    protocols::RequestResponseProtocol,
+};
+use async_trait::async_trait;
+use futures::{
+    AsyncRead,
+    AsyncReadExt,
+    AsyncWriteExt,
+};
+use libp2p::request_response;
+use strum::IntoEnumIterator as _;
+
+use super::{
+    Decode,
+    Encode,
+    Encoder,
+    RequestResponseProtocols,
+};
+
+#[derive(Debug, Clone)]
+pub struct RequestResponseMessageHandler<Codec> {
+    pub(crate) codec: Codec,
+    /// Used for `max_size` parameter when reading Response Message
+    /// Necessary in order to avoid DoS attacks
+    /// Currently the size mostly depends on the max size of the Block
+    pub(crate) max_response_size: usize,
+}
+
+/// Since Postcard does not support async reads or writes out of the box
+/// We prefix Request & Response Messages with the length of the data in bytes
+/// We expect the substream to be properly closed when response channel is dropped.
+/// Since the request protocol used here expects a response, the sender considers this
+/// early close as a protocol violation which results in the connection being closed.
+/// If the substream was not properly closed when dropped, the sender would instead
+/// run into a timeout waiting for the response.
+#[async_trait]
+impl<Codec> request_response::Codec for RequestResponseMessageHandler<Codec>
+where
+    Codec: Encode<RequestMessage, Error = io::Error>
+        + Decode<RequestMessage, Error = io::Error>
+        + Encode<V1ResponseMessage, Error = io::Error>
+        + Decode<V1ResponseMessage, Error = io::Error>
+        + Encode<V2ResponseMessage, Error = io::Error>
+        + Decode<V2ResponseMessage, Error = io::Error>
+        + Send,
+{
+    type Protocol = RequestResponseProtocol;
+    type Request = RequestMessage;
+    type Response = V2ResponseMessage;
+
+    async fn read_request<T>(
+        &mut self,
+        _protocol: &Self::Protocol,
+        socket: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let mut response = Vec::new();
+        socket
+            .take(self.max_response_size as u64)
```

Comment:
This `as` conversion can theoretically error. Personally I'd prefer to use `u64::try_from` explicitly instead, or as suggested before - redefine `max_response_size` as a `u64`.